Feat edge 49 carousel

ComwrapUkReply opened the pull request:
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #<gh-issue-id>

Test URLs:
- Before: https://main--comwrap-whitelabel-eds--comwrapreply.aem.live/
- After: https://feat-edge-49-carousel--comwrap-whitelabel-eds--comwrapreply.aem.page/blocks-demo/carousel
- 
Demo https://author-p24706-e491522.adobeaemcloud.com/content/comwrap-whitelabel-eds/blocks-demo/carousel.html?ref=feat-EDGE-49-carousel

ComwrapUkReply committed:
Refactor carousel component to enhance structure and functionality

- Renamed "Carousel Item" to "Carousel Slide" for consistency across component definitions and models.
- Updated JSON files to reflect the new naming convention and improved structure.
- Enhanced CSS for better responsiveness and added new styles for carousel controls and slide content.
- Introduced JavaScript improvements for autoplay functionality, touch support, and keyboard navigation.
- Added example and README documentation for better usage guidance and accessibility features.

ComwrapUkReply committed:
Refactor carousel component for improved functionality and structure

- Simplified the carousel initialization process by consolidating the logic for autoplay and progress tracking.
- Removed unnecessary transition animations and progress bar updates to streamline performance.
- Enhanced the carousel's slide processing and content handling for better maintainability.
- Updated the event listeners for touch support and autoplay functionality, ensuring a smoother user experience.
- Improved code readability by restructuring functions and removing redundant code.

ComwrapUkReply committed:
Add autoplay option to component models and carousel configuration

- Introduced a new "autoplay" field in both component-models.json and _carousel.json.
- Configured the autoplay option as a boolean with a default value of false, enhancing the carousel's functionality and user control.

ComwrapUkReply committed:
Add autoplay field to carousel component models

- Introduced a new "autoplay" field in both component-models.json and _carousel.json.
- Configured the autoplay option as a boolean, enhancing user control over carousel functionality.

ComwrapUkReply committed:
Add model field to carousel component configuration

- Introduced a new "model" field in the carousel component definition to enhance configuration options.
- Updated the "filter" and "autoplay" fields for improved functionality and user control.

ComwrapUkReply committed:
Refactor carousel component to enhance autoplay functionality

- Removed the "autoplay" field from component models and updated the carousel configuration to streamline autoplay settings.
- Introduced a new "autoplayDelay" field to allow customization of autoplay timing, defaulting to 5 seconds.
- Updated JavaScript logic to retrieve autoplay settings from block data, improving flexibility and user control over carousel behavior.

ComwrapUkReply committed:
Update autoplay field in component models to accept string values

- Changed the "value" of the "autoplay" field from a boolean to an empty string in component-models.json, allowing for more flexible configuration options.

ComwrapUkReply committed:
Update autoplay field in component models to use boolean value

- Changed the "value" of the "autoplay" field from an empty string to false in component-models.json, ensuring consistent boolean configuration for autoplay functionality.

ComwrapUkReply committed:
Update autoplay field in carousel configuration to use empty string value

- Changed the "value" of the "autoplay" field from false to an empty string in both component-models.json and _carousel.json, ensuring consistency in configuration across components.

ComwrapUkReply committed:
Refactor autoplay settings in component models and carousel configuration

- Updated the "autoplay" field in component-models.json and _carousel.json to use a boolean value of false, ensuring consistent configuration.
- Renamed "autoplayDelay" to "autoplayInterval" and changed its default value to 5000 ms, with a description for clarity.
- Added a condition to the "autoplayInterval" field to activate only when autoplay is enabled, enhancing functionality.

ComwrapUkReply committed:
Update carousel configuration to enhance autoplay settings

- Added "valueType" field to the "autoplay" setting in _carousel.json for clarity.
- Changed the default value of "autoplayDelay" from 5 to 5000 milliseconds to align with expected input format.

ComwrapUkReply committed:
Refactor autoplay settings in component models

- Updated the "autoplay" field label to "Enable Autoplay" for clarity.
- Renamed "autoplayInterval" to "autoplayDelay" and changed its label to "Autoplay Delay (seconds)".
- Removed the description and condition from the "autoplayDelay" field to simplify configuration.

ComwrapUkReply committed:
Add pause on hover feature to carousel autoplay settings

- Introduced a new configuration option "pauseOnHover" in component-models.json to allow autoplay to pause when the user hovers over the carousel.
- Updated carousel.js to implement pause on hover functionality, enhancing user experience by providing control over autoplay behavior.
- Adjusted autoplay configuration retrieval to include the new "pauseOnHover" setting, ensuring consistent behavior across components.

ComwrapUkReply committed:
Remove autoplayDelay and pauseOnHover settings from carousel configuration

- Eliminated the "autoplayDelay" and "pauseOnHover" fields from component-models.json and _carousel.json to simplify autoplay settings.
- Updated carousel.js to remove references to the removed settings, streamlining the autoplay functionality and enhancing code clarity.

ComwrapUkReply committed:
Remove autoplay configuration from carousel component models

- Eliminated the "autoplay" field from component-models.json and _carousel.json to simplify carousel settings.
- Updated carousel.js to adjust the autoplay configuration retrieval, now returning an object with autoplay status and interval based on block data attributes.

ComwrapUkReply committed:
Enhance carousel functionality with improved autoplay and animation features

- Updated slide transition duration for smoother animations and adjusted autoplay interval to 6 seconds.
- Implemented auto-looping behavior for slide navigation, allowing seamless transitions between slides.
- Removed autoplay configuration retrieval, simplifying the initialization process and ensuring no autoplay by default.
- Enhanced event listeners and keyboard navigation to support auto-looping, improving user experience.

ComwrapUkReply committed:
Enable autoplay by default for carousel and enhance visibility handling

- Set autoplay to true by default when multiple slides are present, improving user experience.
- Updated stopAutoPlay function to conditionally update the play/pause button state.
- Enhanced Intersection Observer logic to resume autoplay when the carousel becomes visible.
- Cleaned up code formatting for better readability.

ComwrapUkReply committed:
Refactor carousel.js for improved code readability

- Removed unnecessary blank lines to enhance code clarity and maintainability.
- Ensured consistent formatting throughout the file for better visual structure.

ComwrapUkReply committed:
Enhance carousel.js with performance optimizations and new features

- Introduced a debounce function to optimize resize event handling, improving performance during window resizing.
- Added interaction handling to pause autoplay and restart it after user interaction, enhancing user experience.
- Cached DOM queries for slides and dots to improve rendering performance.
- Updated animation handling to utilize requestAnimationFrame for smoother transitions.
- Enhanced event listeners for touch and keyboard navigation, streamlining interaction logic.
- Implemented a cleanup function for better memory management and resource handling.

ComwrapUkReply committed:
Update carousel CSS for improved styling and responsiveness

- Added spacing and adjusted height properties for better layout control.
- Changed media query syntax for improved compatibility.
- Simplified color definitions for consistency and clarity.
- Enhanced hover and background styles for better visual effects.
- Updated comments for clarity on responsive design adjustments.

ComwrapUkReply committed:
Enhance carousel functionality with editor mode support and styling updates

- Implemented event listeners to pause and resume autoplay in editor mode, improving usability for content creators.
- Added a function to visually disable the play/pause button when in editor mode, enhancing user experience.
- Updated CSS to use CSS variables for colors, improving maintainability and consistency across styles.
- Adjusted various height and margin properties for better layout control and responsiveness.

ComwrapUkReply committed:
Enhance carousel.js with improved editor mode handling for autoplay

- Added checks to prevent autoplay when in editor mode, enhancing usability for content creators.
- Updated event listeners to manage autoplay behavior based on editor status, ensuring a smoother editing experience.
- Refined comments for clarity regarding autoplay conditions and editor mode checks.

ComwrapUkReply committed:
Refactor carousel.js to utilize utility function for editor mode checks

- Replaced direct DOM checks for editor mode with a utility function, improving code readability and maintainability.
- Streamlined autoplay logic to enhance performance and user experience when in editor mode.
- Removed redundant code related to editor mode checks, simplifying the overall structure of the carousel functionality.

ComwrapUkReply committed:
Add console log for editor mode status in utils.js

- Introduced a console log statement to output the current state of editor mode, aiding in debugging and development.
- This change enhances visibility into the editor mode functionality, supporting ongoing improvements in related components.

ComwrapUkReply committed:
Update isEditorMode function to check for appContainer instead of editor-app

- Modified the isEditorMode utility function to reference 'appContainer' for detecting editor mode, improving accuracy in mode detection.
- This change aligns with recent updates in the application structure, ensuring consistent behavior across components that rely on editor mode checks.

ComwrapUkReply committed:
Refactor editor mode handling in carousel.js and enhance isEditorMode utility

- Renamed the editor mode observer cleanup function for clarity in carousel.js.
- Improved the isEditorMode function to check both the current document and parent/top frames for the 'appContainer', enhancing accuracy in detecting editor mode.
- Updated the observeEditorMode function to return an array of observers for better management of multiple observers, ensuring robust cleanup capabilities.

ComwrapUkReply committed:
Refactor isEditorMode function to use helper for appContainer checks

- Updated the isEditorMode utility to utilize a helper function for checking the presence of the 'appContainer' class, improving code readability and maintainability.
- Enhanced the function to check both the current document and parent/top frames, ensuring accurate detection of editor mode across different contexts.

ComwrapUkReply committed:
Refactor carousel.js and utils.js for improved robustness and readability

- Enhanced the initializeCarousel function by implementing defensive checks and restructuring state management for better clarity.
- Introduced patterns for slide navigation to streamline the updateCarousel logic.
- Improved the getNodeValue and setBlockItemOptions functions in utils.js with additional defensive checks and error handling.
- Updated the observeEditorMode function to ensure safe callback execution and robust observer management.
- Added comments for clarity and maintainability throughout the code.

ComwrapUkReply committed:
Enhance utils.js with defensive programming and error handling

- Improved getNodeValue and setBlockItemOptions functions with additional defensive checks to ensure valid inputs and prevent runtime errors.
- Updated getBlockChildren to include optional filtering and enhanced error handling for robustness.
- Refined moveClassToTargetedChild with checks for valid block and target elements, ensuring safe class manipulation.
- Enhanced isEditorMode and observeEditorMode functions with better error handling and observer management, improving reliability across different contexts.
- Added comments for clarity and maintainability throughout the code.

ComwrapUkReply committed:
Update Carousel block documentation and enhance content structure

- Expanded the README.md to provide a comprehensive overview of the Carousel block, including usage, variations, authoring guide, developer documentation, accessibility guide, and best practices.
- Improved content structure with a detailed table of contents for easier navigation.
- Added best practices for content authors and developers to optimize carousel usage and performance.
- Included accessibility considerations to ensure compliance with WCAG standards.
- Enhanced examples and customization options for better clarity and usability.

hannahscott16 commented on blocks/carousel/_carousel.json:
```diff
@@ -1,104 +1,77 @@
+          "value": ""
+        },
+        {
+          "component": "richtext",
```
can we make this a title field instead of richtext? 
OR combine the title and text and just have 1 rich text field for both?

ComwrapUkReply replied:
Combined to one rich text!

hannahscott16 commented on blocks/carousel/example.md:
```diff
@@ -0,0 +1,27 @@
+# Carousel Example
```
I don't think we need this file, lets just keep the readme?

hannahscott16 commented on blocks/carousel/README.md:
```diff
@@ -0,0 +1,522 @@
+
+## Usage
+
+To use the Carousel block, create a table in your document with "Carousel" in the first cell, followed by rows containing your slide content. Each row represents one slide.
```
This is talking about document based authoring not UE

ComwrapUkReply replied:
i changed the doc to UE

ComwrapUkReply committed:
Refactor Carousel block for consistency and clarity

- Standardized the title and label casing for "Carousel slide" across component definition and model files to improve consistency.
- Updated the README.md to enhance the structure and clarity of the documentation, including a comprehensive overview and detailed sections for technical architecture and configuration.
- Removed the outdated example.md file to streamline the documentation and focus on relevant content.
- Improved accessibility and usability by refining ARIA attributes and ensuring compliance with best practices.

ComwrapUkReply committed:
Add carousel background color option to component models

- Introduced a new field for "carousel-background-color" in both component models and carousel block definitions.
- Added options for "default" and "dark" background colors to enhance customization capabilities for the carousel component.

ComwrapUkReply committed:
Update carousel component to use select for background color options

- Changed the component type for the carousel background color field from "classes" to "select" in both component models and carousel block definitions.
- Updated the field name to "classes" for consistency across the definitions.

ComwrapUkReply committed:
Add model attribute to Carousel block and update background color options

- Introduced a "model" attribute with the value "carousel" in the Carousel block definition for improved clarity.
- Updated the background color options to have consistent casing for "Default" and "Dark" to enhance user experience and maintain consistency across the component.

ComwrapUkReply committed:
Add model attribute to Carousel block and update background color options

- Introduced a "model" attribute with the value "carousel" in the Carousel block definition for improved clarity.
- Updated the background color options to have consistent casing for "Default" and "Dark" to enhance user experience and maintain consistency across the component.

ComwrapUkReply committed:
Update carousel styles and documentation for improved dark mode support

- Enhanced the carousel CSS to support dark mode with updated background colors and text styles for better visibility.
- Refined button styles in dark mode for improved user experience.
- Expanded the README.md to include a new section on developer guides, providing resources for adding select and multiselect components.

ComwrapUkReply committed:
Refactor carousel styles for improved layout and dark mode support

- Removed outdated mobile styles and consolidated dark mode styles for better clarity and maintainability.
- Updated padding and margin values using CSS variables for consistency.
- Introduced navigation arrows with hover effects for enhanced user interaction.
- Adjusted carousel controls positioning for improved accessibility and usability.
